Sources/NetworkProtection/Recovery/FailureRecoveryHandler.swift

```diff
@@ -0,0 +1,198 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Common
+
+public enum FailureRecoveryStep {
+    public enum ServerHealth {
+        case healthy
+        case unhealthy
+    }
+
+    case started
+    case completed(ServerHealth)
+    case failed(_ error: Error)
+}
+
+protocol FailureRecoveryHandling {
+    func attemptRecovery(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool,
+        updateConfig: @escaping (NetworkProtectionDeviceManagement.GenerateTunnelConfigResult) async throws -> Void
+    ) async
+
+    func stop() async
+}
+
+enum FailureRecoveryError: Error {
+    case noRecoveryNecessary
+}
+
+actor FailureRecoveryHandler: FailureRecoveryHandling {
+
+    struct RetryConfig {
+        let times: Int
+        let initialDelay: TimeInterval
+        let maxDelay: TimeInterval
+        let factor: Double
+
+        static var `default` = RetryConfig(
+            times: 5,
+            initialDelay: .seconds(30),
+            maxDelay: .minutes(5),
+            factor: 2.0
+        )
+    }
+
+    private let deviceManager: NetworkProtectionDeviceManagement
+    private weak var reassertingControl: Reasserting?
+    private let retryConfig: RetryConfig
+    private let eventHandler: (FailureRecoveryStep) -> Void
+
+    private var task: Task<Void, Never>? {
+        willSet {
+            task?.cancel()
+        }
+    }
+
+    init(deviceManager: NetworkProtectionDeviceManagement, reassertingControl: Reasserting, retryConfig: RetryConfig = .default, eventHandler: @escaping (FailureRecoveryStep) -> Void) {
+        self.deviceManager = deviceManager
+        self.reassertingControl = reassertingControl
+        self.retryConfig = retryConfig
+        self.eventHandler = eventHandler
+    }
+
+    func attemptRecovery(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool,
+        updateConfig: @escaping (NetworkProtectionDeviceManagement.GenerateTunnelConfigResult) async throws -> Void
+    ) async {
+        await incrementalPeriodicChecks(retryConfig) { [weak self] in
+            guard let self else { return }
+            await reassertingControl?.startReasserting()
+            eventHandler(.started)
+            do {
+                let result = try await makeRecoveryAttempt(
+                    to: lastConnectedServer,
+                    includedRoutes: includedRoutes,
+                    excludedRoutes: excludedRoutes,
+                    isKillSwitchEnabled: isKillSwitchEnabled
+                )
+                try await updateConfig(result)
+                eventHandler(.completed(.unhealthy))
+                await reassertingControl?.stopReasserting()
```

Comment:
In this snippet, the event is emitted before the `stopReasserting()` is invoked. In the `catch` block, it happens after the `stopReasserting()` call. The order of operations should be consistent, if only for readability.

Reply:
Put it a different way: `reassertingControl?.startReasserting()` is the first thing that should happen, `reassertingControl?.stopReasserting()` is the last. That was my intention here. But I think that can be better expressed with a `defer`[ as Diego suggested](https://github.com/duckduckgo/BrowserServicesKit/pull/786/files#r1580000919)

Reply:
Fair enough, can change that.